spanner/sessionclient_test.go

```diff
@@ -373,3 +375,77 @@ func TestClientIDGenerator(t *testing.T) {
 		}
 	}
 }
+
+func TestMergeCallOptions(t *testing.T) {
+	a := &vkit.CallOptions{
+		CreateSession: []gax.CallOption{
+			gax.WithRetry(func() gax.Retryer {
+				return gax.OnCodes([]codes.Code{
+					codes.Unavailable, codes.DeadlineExceeded,
+				}, gax.Backoff{
+					Initial:    100 * time.Millisecond,
+					Max:        16000 * time.Millisecond,
+					Multiplier: 1.0,
+				})
+			}),
+		},
+		GetSession: []gax.CallOption{
+			gax.WithRetry(func() gax.Retryer {
+				return gax.OnCodes([]codes.Code{
+					codes.Unavailable, codes.DeadlineExceeded,
+				}, gax.Backoff{
+					Initial:    250 * time.Millisecond,
+					Max:        32000 * time.Millisecond,
+					Multiplier: 1.30,
+				})
+			}),
+		},
+	}
+	b := &vkit.CallOptions{
+		CreateSession: []gax.CallOption{
+			gax.WithRetry(func() gax.Retryer {
+				return gax.OnCodes([]codes.Code{
+					codes.Unavailable,
+				}, gax.Backoff{
+					Initial:    250 * time.Millisecond,
+					Max:        32000 * time.Millisecond,
+					Multiplier: 1.30,
+				})
+			}),
+		},
+		BatchCreateSessions: []gax.CallOption{
+			gax.WithRetry(func() gax.Retryer {
+				return gax.OnCodes([]codes.Code{
+					codes.Unavailable,
+				}, gax.Backoff{
+					Initial:    250 * time.Millisecond,
+					Max:        32000 * time.Millisecond,
+					Multiplier: 1.30,
+				})
+			}),
+		}}
+
+	merged := mergeCallOptions(b, a)
+	cs := &gax.CallSettings{}
+	// We can't access the fields of Retryer so we have test the result by
+	// comparing strings.
+	merged.CreateSession[0].Resolve(cs)
+	if got, want := fmt.Sprintf("%v", cs.Retry()), "&{{250000000 32000000000 1.3 0} [14]}"; got != want {
+		t.Fatalf("merged CallOptions is incorrect: got %v, want %v", got, want)
+	}
+
+	merged.CreateSession[1].Resolve(cs)
```

Comment:
I'm not very familiar with the exact working of these options in Go. When exactly would the `merged.CreateSession[1]` setting be used instead of the `merged.CreateSession[0]` setting?

Reply:
https://github.com/googleapis/google-cloud-go/blob/06b6069396a620973802c129f2b58056a079cec5/spanner/apiv1/spanner_client.go#L308-L311

In the generated method, you can see that `opts = [c.CallOptions.CreateSession, opts...]`. So the passed-in opts have higher order of precedence. 

Therefore, `merged.CreateSession[1]` will override `merged.CreateSession[0]`. 

My guess is that it will iterate all `gax.CallOption` in `merged.CreateSession` and call `Resolve` with `gax.CallSettings`. Only the effect of the last one remains. Then, it uses `gax.CallSettings` for calling gRPC.